win32/devdoc/socket_transport_win32.md

```diff
@@ -0,0 +1,296 @@
+```
+
+`socket_transport_create` creates a socket transport.
+
+`socket_transport_create` shall allocate a new `SOCKET_TRANSPORT` object.
+
+`socket_transport_create` shall call `sm_create` to create a sm object.
+
+On any failure `socket_transport_create` shall return `NULL`.
+
+On success `socket_transport_create` shall return `SOCKET_TRANSPORT_HANDLE`.
+
+### socket_transport_destroy
+
+```c
+MOCKABLE_FUNCTION(, void, socket_transport_destroy, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_destroy` destroys all data stored in the `SOCKET_TRANSPORT_HANDLE` object.
+
+If `socket_transport` is `NULL` `socket_transport_destroy` shall do nothing.
+
+`socket_transport_destroy` shall call `sm_destroy` to destroy the sm object.
+
+`socket_transport_destroy` shall free the `SOCKET_TRANSPORT_HANDLE` object.
+
+### socket_transport_connect
+
+```c
+MOCKABLE_FUNCTION(, int, socket_transport_connect, SOCKET_TRANSPORT_HANDLE, socket_transport, const char*, hostname, uint16_t, port, uint32_t, connection_timeout_ms);
+```
+
+`socket_transport_connect` shall connect to a specified endpoint.
+
+If `socket_transport` is `NULL`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If `hostname` is `NULL`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If `port` is `0`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_CLIENT`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `sm_open_begin` to begin the open.
+
+If `sm_open_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `socket` with the params `AF_INET`, `SOCK_STREAM` and `IPPROTO_TCP`.
+
+`socket_transport_connect` shall call `connect_to_endpoint` to connect to the endpoint.
+
+If successful `socket_transport_connect` shall call `sm_open_end` with `true`.
+
+If any failure is encountered, `socket_transport_connect` shall call `sm_open_end` with `false`, fail and return a non-zero value.
+
+### connect_to_endpoint
+
+```c
+static int connect_to_endpoint(SOCKET client_socket, const ADDRINFO* addrInfo, uint32_t timeout_usec)
+```
+
+`connect_to_endpoint` shall call the `connect` API to asynchronously connect to the endpoint.
+
+If the `connect` call fails, `connect_to_endpoint` shall check to `WSAGetLastError` for `WSAEWOULDBLOCK`.
+
+On `WSAEWOULDBLOCK` `connect_to_endpoint` shall call the `select` API with a timeout set in the `zrpc_tcp_client_io_create` function and check the return value:
+
+- If the return is `SOCKET_ERROR`, this indicates a connect and `connect_to_endpoint` shall fail.
+
+- If the return value is 0, this indicates a timeout and `connect_to_endpoint` shall fail.
+
+- Any other value this indicates a possible success and and `connect_to_endpoint` shall test if the socket is writable by calling `FD_ISSET`.
```

Comment:
>FD_ISSET [](http://example.com/codeflow?start=131&length=8)

Is this a windows thingie? I can find it in Linux man pages:

Reply:
Yes this is the way of the Windows.